src/Model/Resolver/CompareProductsResolver.php

```diff
@@ -0,0 +1,181 @@
+
+    /**
+     * @var QuoteIdMaskFactory
+     */
+    private $quoteIdMaskFactory;
+
+    /**
+     * @var ProductDataProvider
+     */
+    private $productDataProvider;
+
+    /**
+     * @var MediaConfig
+     */
+    private $mediaConfig;
+
+    /**
+     * GetCartItems constructor.
+     * @param ListCompare $listCompare
+     * @param StoreManagerInterfaceAlias $storeManager
+     * @param ConfigAlias $catalogConfig
+     * @param VisibilityAlias $catalogProductVisibility
+     * @param Visitor $customerVisitor
+     * @param Session $customerSession
+     * @param QuoteIdMaskFactory $quoteIdMaskFactory
+     * @param ProductDataProvider $productDataProvider
+     * @param MediaConfig $mediaConfig
+     */
+    public function __construct(
+        ListCompare $listCompare,
+        StoreManagerInterfaceAlias $storeManager,
+        ConfigAlias $catalogConfig,
+        VisibilityAlias $catalogProductVisibility,
+        Visitor $customerVisitor,
+        Session $customerSession,
+        QuoteIdMaskFactory $quoteIdMaskFactory,
+        ProductDataProvider $productDataProvider,
+        MediaConfig $mediaConfig
+    ) {
+        $this->listCompare = $listCompare;
+        $this->storeManager = $storeManager;
+        $this->catalogConfig = $catalogConfig;
+        $this->catalogProductVisibility = $catalogProductVisibility;
+        $this->customerVisitor = $customerVisitor;
+        $this->customerSession = $customerSession;
+        $this->quoteIdMaskFactory = $quoteIdMaskFactory;
+        $this->productDataProvider = $productDataProvider;
+        $this->mediaConfig = $mediaConfig;
+    }
+
+    public function resolve(
+        Field $field,
+        $context,
+        ResolveInfo $info,
+        array $value = null,
+        array $args = null
+    ) {
+        $customerId = (int)$context->getUserId();
+        $guestCardId = isset($args['guestCartId']) ? $args['guestCartId'] : null;
+
+        if (!$customerId && !$guestCardId) {
+            return null;
+        }
+
+        $collection = $this->listCompare->getItemCollection();
+
+        if ($customerId) {
+            $collection->setCustomerId($customerId);
+        } elseif ($guestCardId) {
+            $quoteIdMask = $this->quoteIdMaskFactory
+                ->create()
+                ->load($guestCardId, 'masked_id')
+                ->getQuoteId();
+
+            $collection->setVisitorId($quoteIdMask);
+        }
+
+        $store = $this->storeManager->getStore();
+        $collection->setStore($store);
+
+        try {
+            // This loads items but throws undefined method exception which can be ignored
+            $collection->load();
+        } catch (\Exception $e) {}
+
+        $count = $collection->count();
+        $products = [];
+
+        if ($count) {
+            $productIds = $collection->getProductIds();
+
+            foreach ($productIds as $productId) {
+                $item = $this->productDataProvider->getProductDataById((int)$productId);
```

Comment:
This function does not seem legit. Becuase it most probably loads stuff one-by-one per product. This is inefficent.

Reply:
Fixed.